arrow-libs/fx/arrow-fx-coroutines/src/commonMain/kotlin/arrow/fx/coroutines/Resource.kt

```diff
@@ -538,3 +544,26 @@ public fun <A : AutoCloseable> autoCloseable(
 ): Resource<A> = resource {
   autoCloseable(closingDispatcher, autoCloseable)
 }
+
+private fun CompletableJob.completeWith(exitCase: ExitCase) {
+  when (exitCase) {
+    is ExitCase.Cancelled -> cancel(exitCase.exception)
+    is ExitCase.Failure -> completeExceptionally(exitCase.failure)
+    ExitCase.Completed -> complete()
+  }
+}
+
+private inline fun ResourceScope.managedCoroutineScope(coroutineContext: CoroutineContext, jobCreator: (Job?) -> CompletableJob): CoroutineScope {
+  val job = jobCreator(coroutineContext[Job])
+  onRelease { exitCase ->
+    job.completeWith(exitCase)
+    job.join()
+  }
+  return CoroutineScope(coroutineContext + job)
+}
+
+public fun ResourceScope.managedSupervisorScope(coroutineContext: CoroutineContext = EmptyCoroutineContext): CoroutineScope =
```

Comment:
Just a suggestion. I'm worried this breaks structured concurrency because it doesn't automatically inherit the current coroutine scope. Maybe a contextual version of it could be nice to have?

Reply:
I intended this to be analogous to the `CoroutineScope` "constructor" which doesn't automatically inherit context, but with its lifecycle coupled to the containing resource scope. 

Given the nature of resource scope, structured concurrency should be maintained even without the job inheritance. 

One could always pass`currentCoroutineContext()` if truly desired.